Cross-match PGIR alert stream with public ZTF data

dekishalay opened the pull request:
This PR aims to add ZTF crossmatches (public data only) for all alerts incoming in the PGIR stream. We'd like to set up scanning to include PGIR sources that have ZTF crossmatches. 

Added ZTF_alerts to crossmatch list for PGIR in config.defaults.yaml. Also included filter to include only programid =1.

dmitryduev reviewed (changes requested):
Thanks, @dekishalay. There are two issues with this implementation that I can see: 
- if done this way, the cross-match would suck in and save to `PGIR_alerts_aux` the basic info about potentially many thousands of ZTF alerts, which is a waste of space and likely useless.
- You would get the x-match for only the first time an `objectId` is saved to `PGIR_alerts_aux`, so no new ZTF alert x-matches would propagate there. [This mechanism is meant for static catalogs , and this first-time-save-only mechanism is a deliberate optimization].

Instead, I would recommend implementing a custom filter, in a way similar to how CLU x-match is dealt with, see https://github.com/dmitryduev/kowalski/blob/master/kowalski/alert_broker_pgir.py#L94.
In particular, you may wish to grab only some basic info about the ZTF alerts/objectIds, perhaps only for a limited number of those, or only for the most recent ones -- depending on what will be guiding your decision making when scanning the candidates. If you'd want to run it for every PGIR alert (i.e. not just when an objectId is first encountered by K), the underlying queries have to be efficient.

dmitryduev commented:
Also, looks like your pre-commit hook is not working -- linting failed on the CI, please check.

dekishalay commented:
Thanks @dmitryduev for the review. I examined the code in alert_broker_pgir.py, and it appears that even the CLU crossmatch is done only once? The external (static) and CLU crossmatches appear to happen inside
       
 if (
            alert_worker.mongo.db[alert_worker.collection_alerts_aux].count_documents(
                {"_id": object_id}, limit=1
            )
            == 0
        ):

which would suggest a one-time crossmatch? Do you suggest adding the ZTF crossmatch outside this if construct?

dmitryduev commented:
> Do you suggest adding the ZTF crossmatch outside this if construct?

Yes, exactly. And as you can see, it should be performant since you'd be running it on each and every alert.

dekishalay commented:
Hi @dmitryduev, okay I've now added a new function in alert_broker.py to find the most recent ZTF alert (made the function more generic than just for ZTF alerts). Also added adding the ZTF alert to aux in alert_broker_pgir.py

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+    ) -> dict:
+        """
+        Run cross-match with a live catalog that is constantly updated
+        Only searches for the most entry to keep it efficient
```
```suggestion
        Only searches for the most recent entry to keep it efficient
```

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+            ra_geojson = ra - 180.0
+            dec_geojson = dec
+
+            catalog_filter = {"candidate.programid": {"$eq": 1}}
```
Why only public data? I am afraid this has to be implemented differently: e.g., grab and store the latest per programid, then filter downstream? Or else, is it true that anyone who has access to PGIR also has access to ZTF programid=3? @doccosmos? That would simplify this by a lot.

dekishalay replied:
The PGIR team includes several members outside ZTF, which is why we want only programid=1 data by default. Only a few members in PGIR have access to programid=3 in ZTF. Why would this programid filter be a problem?

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+        xmatches = dict()
+
+        # cone search radius in arcsec:
+        cone_search_radius_live = 8.0
```
Why 8"? I would probably make this configurable and move this over to config.

dekishalay replied:
8.0" is the typical PSF size for PGIR. Okay to move it to PGIR to enable re-use of the code for other types of cross-matches

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+            catalog_filter = {"candidate.programid": {"$eq": 1}}
+            catalog_sort = {"candidate.jd": -1}
+            catalog_projection = {
+                "objectId": 1,
```
```suggestion
                "objectId": 1,
                "candid": 1,
```

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+                }
+            }
+            # Just find the most recent alert within the crossmatch radius, if any
+            live_alert = self.mongo.db[live_stream].find_one(
```
```suggestion
            latest_alert = list(self.mongo.db[live_stream].find(
```

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+            dec_geojson = dec
+
+            catalog_filter = {"candidate.programid": {"$eq": 1}}
+            catalog_sort = {"candidate.jd": -1}
```
```suggestion
            catalog_sort = [("candidate.jd", -1)]
```

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+            # Just find the most recent alert within the crossmatch radius, if any
+            live_alert = self.mongo.db[live_stream].find_one(
+                {**object_position_query, **catalog_filter},
+                projection={**catalog_projection},
```
Projection only works with find, find_one return the full doc.

dekishalay replied:
Okay, thanks for the clarification

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+            live_alert = self.mongo.db[live_stream].find_one(
+                {**object_position_query, **catalog_filter},
+                projection={**catalog_projection},
+                sort={**catalog_sort},
```
```suggestion
                sort=catalog_sort,
                limit=1,
```

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+                {**object_position_query, **catalog_filter},
+                projection={**catalog_projection},
+                sort={**catalog_sort},
+            )
```
```suggestion
            ))
```

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -861,6 +861,73 @@ def alert_filter__xmatch_clu(
+            )
+
+            # Check if any result was found (live_alert is not null)
+            if live_alert:
+                xmatches[live_stream] = [live_alert]
+            else:
+                xmatches[live_stream] = []
```
```suggestion
            xmatches[live_stream] = latest_alert if latest_alert else []
```

dmitryduev commented on kowalski/alert_broker_pgir.py:
```diff
@@ -120,6 +129,21 @@ def process_alert(alert: Mapping, topic: str):
+            ):
+                alert_worker.mongo.db[alert_worker.collection_alerts_aux].update_one(
+                    {"_id": object_id},
+                    {"$addToSet": {"ZTF_alerts": {"$each": xmatches_ztf}}},
```
Did you mean
```suggestion
                    {"$set": {"ZTF_alerts": xmatches_ztf}},
```
? What you have here will add the latest ZTF alert to this list (treating it as a set to avoid duplicates).

dekishalay replied:
Right, your suggestion looks better since we would avoid adding new alerts from the same ZTF name repeatedly. Okay to include only the latest ZTF alert.

dmitryduev reviewed (changes requested):
Thanks @dekishalay. Please see my comments. 
I think this will need to tested at scale on staging before rolling out to production.

dmitryduev commented:
@dekishalay could you please fix linting and commit it?

dekishalay commented:
@dmitryduev Merged the base branch and fixed linting for review

dmitryduev commented on kowalski/alert_broker.py:
```diff
@@ -866,6 +866,74 @@ def alert_filter__xmatch_clu(
 
         return xmatches
 
+    def alert_filter__xmatch_live_stream(
```
Now that I think about this, your implementation is too PGIR->ZTF specific, so does not belong in the base class (`AlertWorker`), but should go to `PGIRAlertWorker` subclass in `alert_broker_pgir.py`. I would also rename it into ...__xmatch_ztf_alerts and ditch "live_stream" in the func args as your filter, projection, and sort options are very ZTF-specific (like candidate.programid=1).

dekishalay replied:
Thanks @dmitryduev for the suggestion. I've now implemented the change in the next commit.

dmitryduev reviewed (changes requested):
Thanks @dekishalay, please see my comment. 

I would also like to ask you to test your query out with penquins/Kowalski - this would be more representative than a test on staging. I.e.: iterate over, say 1k PGIR alerts from a night, get their ra/decs, query K with a `cone_search` query from `penquins` with these additional filters and a kwargs.limit=1, and see how long it takes.

dekishalay commented:
> your

I also ran a test on K to understand how much time it takes to run these crossmatches. Selecting random 1K PGIR alerts from several nights, I ran the same filter and sort on penquins to confirm that the operations take between 1 - 4 seconds, with a median of ~ 2 seconds for 1 K alerts. 

Is this fast enough @dmitryduev ?

dmitryduev reviewed (approved):
A couple last comments and it's good to go in + please rebase off the latest master, that should the failing test.

The index I'm building on the prod db will speed things up a bit, but not dramatically, so should be good enough already:

```JSON
{
  "coordinates.radec_geojson": "2dsphere",
  "candidate.programid": 1,
  "candidate.jd": 1
}
```

dekishalay commented:
Thanks @dmitryduev . Committed the suggestions and tests appear to pass. Ready to merge?

dmitryduev commented:
Thanks, @dekishalay. Yes, I think it can go in now. The index I was creating is ready so could you please rerun your query timing test once again in the mean time?

dekishalay commented:
Thanks! I tried the test again, and it appears to be hovering between 1 and 3 seconds now.
